specutils/analysis/resample.py

```diff
@@ -0,0 +1,219 @@
+    delta_fin = fin_lamb[1] - fin_lamb[0]
+
+    n_orig_lamb = len(orig_lamb)
+    n_fin_lamb = len(fin_lamb)
+
+    # Lower bin and upper bin edges
+    orig_low = orig_lamb - delta_orig * 0.5
+    orig_upp = orig_lamb + delta_orig * 0.5
+    fin_low = fin_lamb - delta_fin * 0.5
+    fin_upp = fin_lamb + delta_fin * 0.5
+
+    # Create re-sampling matrix
+    resamp_mat = np.zeros(shape=(n_fin_lamb, n_orig_lamb))
+
+    for i in range(n_fin_lamb):
+        # Calculate the contribution of each original bin to the
+        # resampled bin
+        l_inf = np.where(orig_low > fin_low[i], orig_low, fin_low[i])
+        l_sup = np.where(orig_upp < fin_upp[i], orig_upp, fin_upp[i])
+
+        # Interval overlap of each original bin for current resampled
+        # bin; negatives clipped
+        dl = (l_sup - l_inf).clip(0)
+
+        # This will only happen at the edges of lorig.
+        # Discard resampled bin if it's not fully covered (> 99%) by the
+        #  original bin -- only happens at the edges of the original bins
+        if 0 < dl.sum() < 0.99 * delta_fin:
+            dl = 0 * orig_lamb
+
+        resamp_mat[i, :] = dl
+
+    resamp_mat = resamp_mat / delta_fin
+
+    return resamp_mat
+
+
+def _nonuniform_matrix(orig_lamb, fin_lamb, extrapolate=False):
+    """
+    Compute re-sampling matrix R_o2r, useful to convert a spectrum sampled at
+    wavelengths lorig to a new grid lresamp. Here, there is no necessity to
+    have constant gris as on :func:`_uniform_matrix`. This is adapted from code
+    created by the SEAGal Group.
+
+    .. warning:: lorig and lresam MUST be on ascending order!
+
+    Parameters
+    ----------
+    orig_lamb : array_like
+        Original spectrum lambda array.
+    fin_lamb : array_like
+        Spectrum lambda array in which the spectrum should be sampled.
+    extrapolate : boolean, optional
+        Extrapolate values, i.e. values for fin_lamb < orig_lamb[0] are set to
+        match orig_lamb[0] and values for fin_lamb > orig_lamb[-1] are set to
+        match orig_lamb[-1].
+
+    Returns
+    -------
+    resample_mat : ndarray
+        Resample matrix.
+    """
+    matrix = np.zeros(shape=(len(fin_lamb), len(orig_lamb)))
+
+    # Define lambda ranges (low, upp) for original and resampled.
+    lo_low = np.zeros(len(orig_lamb))
+    lo_low[1:] = (orig_lamb[1:] + orig_lamb[:-1]) / 2
+    lo_low[0] = orig_lamb[0] - (orig_lamb[1] - orig_lamb[0]) / 2
+
+    lo_upp = np.zeros(len(orig_lamb))
+    lo_upp[:-1] = lo_low[1:]
+    lo_upp[-1] = orig_lamb[-1] + (orig_lamb[-1] - orig_lamb[-2]) / 2
+
+    lr_low = np.zeros(len(fin_lamb))
+    lr_low[1:] = (fin_lamb[1:] + fin_lamb[:-1]) / 2
+    lr_low[0] = fin_lamb[0] - (fin_lamb[1] - fin_lamb[0]) / 2
+
+    lr_upp = np.zeros(len(fin_lamb))
+    lr_upp[:-1] = lr_low[1:]
+    lr_upp[-1] = fin_lamb[-1] + (fin_lamb[-1] - fin_lamb[-2]) / 2
+
+    # Iterate over resampled fin_lamb vector
+    for i in range(len(fin_lamb)):
```

Comment:
could you use a >=2 character iteration variable here?

Reply:
How do you mean? Only iterate if the final grid has more than a single point?